File: algo_problems/q861-880/q873.java
class Solution {
    public int lenLongestFibSubseq(int[] A) {
        Map<Integer, Integer> indexMap = new HashMap();
        for (int i = 0; i < A.length; i ++){
        	indexMap.put(A[i], i);
        }
        int result = 0;
        Map<Integer, Integer> longestMap = new HashMap();
        int n = A.length;
        for (int i = 0; i<n; i++){
        	for (int j = 0; j<i; j++){
        		int t = A[i] - A[j];
        		int k = indexMap.getOrDefault(t, -1);
        		if (k != -1 && j < k){
        			// longestMap[k, i] = lonmap[j, k] + 1
        			int cand = longestMap.getOrDefault(j * n + k, 2) + 1;
        			longestMap.put(k * n + i, cand);
        			result = Math.max(result, cand);
        		}
        	}
        }
        return result;

    }
}

File: algo_problems/q861-880/q873.py
import collections


class Solution:
    def lenLongestFibSubseq(self, A):
        """
        :type A: List[int]
        :rtype: int
        """
        table = collections.defaultdict(lambda: 2)
        indexDict = {v: k for k, v in enumerate(A)}
        result = 0
        for i, v in enumerate(A):
            for j in range(i):
                t = v - A[j]
                k = indexDict.get(t, None)
                if k is not None and j < k:
                    cand = table[k, i] = table[j, k] + 1
                    result = max(result, cand)
        return result


if __name__ == '__main__':
    print(Solution().lenLongestFibSubseq([1, 2, 3, 4, 5, 6, 7, 8]))
